### PR TITLE
Make UGI current user the default owner of Hive db&tbl

### PR DESCRIPTION
**Pull request by haizhou-zhao**
Following up PR on https://github.com/apache/iceberg/pull/6045

Change Summary:
1. [HiveCatalog] Change default owner of Hive db&tbl from OS username to UGI current user, this impacts create, alter and remove operations of db&table in HiveCatalog
2. Relevant unit test changes

### CODE REVIEW

**Comment by haizhou-zhao** [comment]
@danielcweeks @gaborkaszab @szehon-ho ^

Feel free to invite other folks to review.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 1]
```diff
@@ -567,8 +569,13 @@ Database convertToDatabase(Namespace namespace, Map<String, String> meta) {
         });
 
     if (database.getOwnerName() == null) {
-      database.setOwnerName(System.getProperty("user.name"));
-      database.setOwnerType(PrincipalType.USER);
+      try {
+        database.setOwnerName(UserGroupInformation.getCurrentUser().getUserName());
+        database.setOwnerType(PrincipalType.USER);
+      } catch (IOException e) {
+        throw new RuntimeException(
```
Comment (szehon-ho) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:576]: I think this was also discussed in https://github.com/apache/iceberg/pull/6045#discussion_r1004158438.  Should we just warn and fallback?  @gaborkaszab @danielcweeks 

by the way, if we throw we have our own UnchekcedIOException

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 2]
```diff
@@ -470,6 +481,22 @@ private void setHmsTableParameters(
     // remove any props from HMS that are no longer present in Iceberg table props
     obsoleteProps.forEach(parameters::remove);
 
+    if (metadata.property(HiveCatalog.HMS_TABLE_OWNER, null) != null) {
```
Comment (szehon-ho) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:484]: Looks like this code is for the update case and is going a bit beyond original behavior (which was for create case).  

cc @gaborkaszab for thoughts

And does it align with database/namespace update?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 1 above)
Comment (danielcweeks) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:576]: Looking through the `UGI` path, it looks like the only likely failure is that security (e.g. kerberos) is enabled, but something is misconfigured or the user isn't logged in.  It probably makes sense to just throw here, but we should use the java built-in `UncheckedIOException` since we have an IO cause (Iceberg's `RuntimeIOException` is probably best for when we don't have an IO cause).

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (gaborkaszab) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:484]: Agree @szehon-ho that this seems a different use-case for me. Might worth covering it in a separate PR.

Additionally, this function is for setting the table parameters and with this change it would also have a side effect of setting the tbl's owner that is not really a table property. I would feel this is a bit off in my opinion.

And anyway, when this code was run for a "create table" scenario then newHmsTable() would have already set the table owner so the below line would be redundant.

But first of all as this is for the update scenario I'd say that this deserve a separate PR because this one is for changing the defaults during table creation.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 1 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:576]: ack taken

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 3]
```diff
@@ -422,20 +424,29 @@ private Table newHmsTable(TableMetadata metadata) {
...
-            (int) currentTimeMillis / 1000,
-            Integer.MAX_VALUE,
-            null,
-            Collections.emptyList(),
-            Maps.newHashMap(),
-            null,
-            null,
-            TableType.EXTERNAL_TABLE.toString());
+    Table newTable;
+    try {
```
[the remark this answers is not in the document]
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:428]: make sense, taken

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:484]: cool, I can separate it out into a separate PR. I hope to discuss whether this feature is needed. Basically, what I'm trying to achieve here is users can tweak ownership by setting iceberg table properties. And should users remove the owner properties, then it falls back to default owner.

[the remark this answers is not in the document]
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:484]: ack

Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:484]: I put it there cause I was thinking to unify the ownership behavior between database and table.

**Comment by haizhou-zhao** [comment]
Hey folks, comments from last round of review all taken and implemented.

Specifically, on one comment:
@gaborkaszab @szehon-ho I removed support for changing ownership for tables from this PR. But before I committed to create a follow up PR on that, I want to double check and make sure if you guys feel these worth implementing (which are something we do not have today):
1. When user remove ownership on a table, revert the owner of the table to default owner (current UGI)
2. When user alter the ownership on a table, change the owner of table accordingly

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 4]
```diff
@@ -422,11 +425,21 @@ private Table newHmsTable(TableMetadata metadata) {
     Preconditions.checkNotNull(metadata, "'metadata' parameter can't be null");
     final long currentTimeMillis = System.currentTimeMillis();
 
+    String defaultUser;
+    try {
```
Comment (gaborkaszab) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:429]: Actually there is one thing: I think in case the HMS_TABLE_OWNER is set we don't have to query the UGI.getCurrentUser() because it won't be used anyway.

**Comment by gaborkaszab** [comment]
> Hey folks, comments from last round of review all taken and implemented.
> 
> Specifically, on one comment: @gaborkaszab @szehon-ho I removed support for changing ownership for tables from this PR. But before I committed to create a follow up PR on that, I want to double check and make sure if you guys feel these worth implementing (which are something we do not have today):
> 
> 1. When user remove ownership on a table, revert the owner of the table to default owner (current UGI)
> 2. When user alter the ownership on a table, change the owner of table accordingly

I think it does make sense to have a support for altering ownership. Impala has "ALTER TABLE xy SET OWNER" for this purpose, and I guess Hive also has something similar.
Yes, the use-cases 1) and 2) you mention make sense for me.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 4 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:429]: Ack, taken

**Comment by haizhou-zhao** [comment]
> > Hey folks, comments from last round of review all taken and implemented.
> > Specifically, on one comment: @gaborkaszab @szehon-ho I removed support for changing ownership for tables from this PR. But before I committed to create a follow up PR on that, I want to double check and make sure if you guys feel these worth implementing (which are something we do not have today):
> > 
> > 1. When user remove ownership on a table, revert the owner of the table to default owner (current UGI)
> > 2. When user alter the ownership on a table, change the owner of table accordingly
> 
> I think it does make sense to have a support for altering ownership. Impala has "ALTER TABLE xy SET OWNER" for this purpose, and I guess Hive also has something similar. Yes, the use-cases 1) and 2) you mention make sense for me.

Thank you @gaborkaszab , I create a follow up PR for those use cases.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 5]
```diff
@@ -567,8 +570,13 @@ Database convertToDatabase(Namespace namespace, Map<String, String> meta) {
...
 
     if (database.getOwnerName() == null) {
-      database.setOwnerName(System.getProperty("user.name"));
-      database.setOwnerType(PrincipalType.USER);
+      try {
+        database.setOwnerName(UserGroupInformation.getCurrentUser().getUserName());
+        database.setOwnerType(PrincipalType.USER);
+      } catch (IOException e) {
+        throw new UncheckedIOException(
+            String.format("Fail to obtain default (UGI) user for database %s", database), e);
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:578]: I think this should fall back to the previous behavior. This should be a warning, not an exception.

Also, I think the message could be more clear: "Failed to fetch Hadoop user". No need to log the database, it's unrelated to what failed.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 6]
```diff
@@ -422,11 +425,23 @@ private Table newHmsTable(TableMetadata metadata) {
     Preconditions.checkNotNull(metadata, "'metadata' parameter can't be null");
     final long currentTimeMillis = System.currentTimeMillis();
 
+    String owner = metadata.properties().get(HiveCatalog.HMS_TABLE_OWNER);
+    if (owner == null) {
+      try {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:430]: Since this is used in two places, can you make it a static method in a Hadoop util class?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java [hunk 7]
```diff
@@ -246,23 +248,33 @@ public void testReplaceTxnBuilder() throws Exception {
...
-      Assert.assertEquals(owner, hmsTable.getOwner());
-      Map<String, String> hmsTableParams = hmsTable.getParameters();
-      Assert.assertFalse(hmsTableParams.containsKey(HiveCatalog.HMS_TABLE_OWNER));
+      Schema schema = getTestSchema();
+      PartitionSpec spec = PartitionSpec.builderFor(schema).bucket("data", 16).build();
+      String location = temp.newFolder(tbl).toString();
+      catalog.createTable(TableIdentifier.of(db, tbl), schema, spec, location, prop);
+      org.apache.hadoop.hive.metastore.api.Table hmsTable = metastoreClient.getTable(db, tbl);
+      Assert.assertEquals(expectedOwner, hmsTable.getOwner());
+      Assert.assertFalse(hmsTable.getParameters().containsKey(HiveCatalog.HMS_TABLE_OWNER));
```
Comment (rdblue) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:273]: Please roll back unnecessary changes to this code block. For example, removing `hmsTableParams`. We want to minimize changes.

Also, this can throw Exception. There is no need for try/catch.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 5 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:578]: Agree on comments for error message. On the other point, I assume you are suggesting if we cannot obtain UGI identity, then we should use OS username like before. I'd like to put more discussion on it.

This PR is created because OS username (System.getProperty("user.name")) is not a good representation of who the owner of this process is when interacting with Hive Metastore (@danielcweeks @szehon-ho @gaborkaszab correct me if needed). Hive Metastore and its accompanying execution engines (HQL or Spark) usually recognize the Kerberos principal as identity, hence Hadoop's UGI is chosen over OS username.

When UGI fail to get the current identity, that means the user do not have the correct keytab or other credentials to authenticate with Kerberos server. If we fall back to OS username, then the outcome is we are swallowing a user issue, and setting the owner to some user unrecognizable by Hive ecosystem. If user didn't set up their Kerberos credentials properly, then I think it's the user's issue, and the cleanest way is to throw the error back at user, instead of swallowing the error and try to do something for user.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 6 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:430]: Ack

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java (hunk 7 above)
Comment (haizhou-zhao) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:273]: Ack

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import org.apache.hadoop.security.UserGroupInformation;
+
+public class HiveHadoopUtil {
+
+  private HiveHadoopUtil() {}
+
+  public static String getUserName() {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java:29]: How about `currentUser` instead of `getUserName`? The verb "get" is not valuable and so Iceberg doesn't use it. Instead, we use a more specific word or omit it. In this case, I think `currentUser` makes sense.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    try {
+      return UserGroupInformation.getCurrentUser().getUserName();
+    } catch (IOException e) {
+      throw new UncheckedIOException("Fail to obtain Hadoop UGI user", e);
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java:33]: In what cases is `IOException` thrown? Should this fall back to the old behavior and return `System.getProperty("user.name")` instead? I don't like that we have now introduced a new error case based on Hadoop internals.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java (hunk 8 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java]: changed in the newest commit.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java (hunk 9 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java]: Falling back to os username in the newest commit.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveHadoopUtil.class);
+
+  private HiveHadoopUtil() {}
+
+  public static String currentUser() {
+    try {
+      return UserGroupInformation.getCurrentUser().getUserName();
+    } catch (IOException e) {
+      LOG.warn("Fail to obtain Hadoop UGI user", e);
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java:36]: Minor: to fit with conventions, I think the error message should be `Failed to get Hadoop user`.
* Error messages use past tense to describe what was attempted
* "Obtain" is no more specific than "get", so prefer the simpler word
* `UGI` is an internal Hadoop term that is more confusing than helpful

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java [hunk 11]
```diff
@@ -247,17 +249,28 @@ public void testReplaceTxnBuilder() throws Exception {
 
   @Test
   public void testCreateTableWithOwner() throws Exception {
+    createTableAndVerifyOwner(
```
Comment (rdblue) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:252]: Why are these mixed into a single test method rather than two test methods? I think this makes it harder to follow

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java (hunk 10 above)
Comment (haizhou-zhao) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveHadoopUtil.java]: Adopted in the latest commit.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @haizhou-zhao!
